net/ipfrag/ipfrag.c

```diff
@@ -0,0 +1,1257 @@
+      if (iob_tryadd_queue(iob, fragq) < 0)
+        {
+          iob_free(iob);
+          iob = NULL;
+        }
+    }
+
+  return iob;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: ip_frag_remnode
+ *
+ * Description:
+ *   free ip_fragsnode_s
+ *
+ * Returned Value:
+ *   I/O buffer count of this node
+ *
+ ****************************************************************************/
+
+uint32_t ip_frag_remnode(FAR struct ip_fragsnode_s *node)
+{
+  g_bufoccupy -= node->bufcnt;
+  assert(g_bufoccupy < CONFIG_IOB_NBUFFERS);
+
+  sq_rem((FAR sq_entry_t *)node, &g_assemblyhead_ipid);
+  sq_rem((FAR sq_entry_t *)&node->flinkat, &g_assemblyhead_time);
+
+  return node->bufcnt;
+}
+
+/****************************************************************************
+ * Name: ip_fragin_enqueue
+ *
+ * Description:
+ *   Enqueue one fragment.
+ *   All fragments belonging to one IP frame are organized in a linked list
+ *   form, that is a ip_fragsnode_s node. All ip_fragsnode_s nodes are also
+ *   organized in an upper-level linked list.
+ *
+ * Returned Value:
+ *   Whether queue is empty before enqueue the new node
+ *
+ ****************************************************************************/
+
+bool ip_fragin_enqueue(FAR struct net_driver_s *dev,
+                       FAR struct ip_fraglink_s *curfraglink)
+{
+  FAR struct ip_fragsnode_s *node;
+  FAR sq_entry_t         *entry;
+  FAR sq_entry_t         *entrylast = NULL;
+  bool                    empty;
+
+  /* The linked list is ordered by IPID value, walk through it and try to
+   * find a node that has the same IPID value, otherwise need to create a
+   * new node and insert it into the linked list.
+   */
+
+  entry = sq_peek(&g_assemblyhead_ipid);
+  empty = (entry == NULL) ? true : false;
+
+  while (entry)
+    {
+      node = (struct ip_fragsnode_s *)entry;
+
+      if (dev == node->dev && curfraglink->ipid <= node->ipid)
+        {
+          break;
+        }
+
+      entrylast = entry;
+      entry = sq_next(entry);
+    }
+
+  node = (struct ip_fragsnode_s *)entry;
+
+  if (node != NULL && curfraglink->ipid == node->ipid)
+    {
+      FAR struct ip_fraglink_s *fraglink;
+      FAR struct ip_fraglink_s *lastlink = NULL;
+
+      /* Found a previously created ip_fragsnode_s, insert this new
+       * ip_fraglink_s to the subchain of this node.
+       */
+
+      fraglink = node->frags;
+
+      /* A ip_fragsnode_s must have a ip_fraglink_s because we allocate a new
+       * ip_fraglink_s when caching a new ip_fraglink_s with a new IPID
+       */
+
+      while (fraglink)
+        {
+          /* The fragment list is ordered by fragment offset value */
+
+          if (curfraglink->fragoff <= fraglink->fragoff)
+            {
+              break;
```

Comment:
Please include an Info/Warn message saying the that fragment is out of order

Reply:
> Please include an Info/Warn message saying the that fragment is out of order

Because these packets are received from the network, the sequence cannot be guaranteed due to network fluctuations，this is a normal phenomenon, so i don't think warning message is necessary.